### ISSUE TITLE
Warp mouse on xwayland can be fixed

### ISSUE THREAD

**Issue by sparklost** [libsdl-org/SDL#9539]
Wayland does not allow applications to warp mouse, but application ran through Xwayland can. [Proof](https://gitlab.freedesktop.org/xorg/xserver/-/blob/xwayland-21.1/hw/xwayland/xwayland-cursor.c#L410).

But applications running SDL2 can't warp mouse (Coming from pygame-ce [issue](https://github.com/pygame-community/pygame-ce/issues/2799)).
However, Blender, that also uses SDL2 can warp. 
After some investigation, I found how they fixed it, in one [bug report](https://projects.blender.org/blender/blender/issues/53004#issuecomment-551561):
>The issue is that the cursor cannot be warped as long as it is visible. The way things work currently is, if a cursor is hidden and it is warped, we lie to X11 clients and say it is actually warped. While doing so, we lock the pointer via the available Wayland protocols. When the cursor is made visible or that for some other reason the lock is broken (for example the user Alt-tab:ed away), we stop lying and where the visible cursor is displayed is the position we tell the clients about. This works for the majority of use cases we have run into (including Blender uses, excluding the one discussed in this report).
We handle confinement in a more direct way.
In other words, if the patch implements a behavior where the pointer cursor is warped while visible, it wont work. If the pointer cursor is confined using X11 protocol, or hidden then warped, and it doesn't work, it could very much be an Xwayland issue.

So, to get this working on on Xwayland, mouse cursor has to be hidden, warped, and shown again.

**Comment by Starbuck5** [comment]
@Kontrabant thanks for fixing this so quickly!

Do you think this is worth porting to SDL2?

**Comment by kjliew** [comment]
Patch for SDL2-2.30.2

Simplify to do this regardless of Xwayland for consistency.

```
diff -ru orig/SDL2-2.30.2/src/video/x11/SDL_x11mouse.c src/SDL2-2.30.2/src/video/x11/SDL_x11mouse.c
--- orig/SDL2-2.30.2/src/video/x11/SDL_x11mouse.c
+++ src/SDL2-2.30.2/src/video/x11/SDL_x11mouse.c
@@ -317,6 +317,10 @@
 {
     SDL_VideoData *videodata = (SDL_VideoData *)SDL_GetVideoDevice()->driverdata;
     Display *display = videodata->display;
+    SDL_Mouse *mouse = SDL_GetMouse();
+    SDL_bool cursor_shown = (mouse)? mouse->cursor_shown:SDL_FALSE;
+    if (cursor_shown)
+        X11_ShowCursor(NULL);
 #ifdef SDL_VIDEO_DRIVER_X11_XINPUT2
     int deviceid = 0;
     if (X11_Xinput2IsInitialized()) {
@@ -335,6 +339,8 @@
     {
         X11_XWarpPointer(display, None, xwindow, 0, 0, 0, 0, x, y);
     }
+    if (cursor_shown)
+        X11_ShowCursor(SDL_GetCursor());
     X11_XSync(display, False);
     videodata->global_mouse_changed = SDL_TRUE;
 }
```

**Comment by ankith26** [comment]
Just putting my 2 cents, IMHO this is worth backporting to SDL2

**Comment by slouken** [comment]
And merged to release-2.30.x in c39f86c28

**Comment by ankith26** [comment]
Thanks! :partying_face:
